File: Assets/Resources/LuaScripts/Tests/System.lua
return function()
    local s = System

    local function assertType(name, val, t)
        if type(val) ~= t then
            error(name .. " expected " .. t .. ", got " .. type(val))
        end
    end

    assertType("clock", s.clock(), "number")
    assertType("time", s.time(), "number")
    assertType("timeMillis", s.timeMillis(), "number")

    assertType("date", s.date(), "string")
    assertType("utcDate", s.utcDate(), "string")

    local env = s.getenv("PATH")
    if env ~= nil then
        assertType("getenv", env, "string")
    end

    assertType("platform", s.platform(), "string")
    assertType("os", s.os(), "string")
    assertType("deviceName", s.deviceName(), "string")
    assertType("deviceModel", s.deviceModel(), "string")

    assertType("processorCount", s.processorCount(), "number")
    assertType("systemMemoryMB", s.systemMemoryMB(), "number")

    assertType("documentsPath", s.documentsPath(), "string")
    assertType("persistentPath", s.persistentPath(), "string")
    assertType("tempPath", s.tempPath(), "string")

    assertType("frameCount", s.frameCount(), "number")
    assertType("deltaTime", s.deltaTime(), "number")

    return 0
end

File: Assets/Scripts/Services/system.cs
using MoonSharp.Interpreter;
using System;
using UnityEngine;

public class SystemService : LuaService
{
    public override void Register(Script script)
    {
        var sys = new Table(script);

        sys["clock"] = (Func<double>)(() => Time.realtimeSinceStartupAsDouble);
        sys["time"] = (Func<long>)(() => DateTimeOffset.UtcNow.ToUnixTimeSeconds());
        sys["timeMillis"] = (Func<long>)(() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());

        sys["date"] = (Func<string, string>)((format) =>
        {
            var dt = DateTime.Now;
            if (string.IsNullOrEmpty(format) || format == "*t")
                return dt.ToString("yyyy-MM-dd HH:mm:ss");
            return dt.ToString(format);
        });

        sys["utcDate"] = (Func<string, string>)((format) =>
        {
            var dt = DateTime.UtcNow;
            if (string.IsNullOrEmpty(format) || format == "*t")
                return dt.ToString("yyyy-MM-dd HH:mm:ss");
            return dt.ToString(format);
        });

        sys["getenv"] = (Func<string, string>)((key) => Environment.GetEnvironmentVariable(key));

        sys["platform"] = (Func<string>)(() => Application.platform.ToString());
        sys["os"] = (Func<string>)(() => SystemInfo.operatingSystem);
        sys["deviceName"] = (Func<string>)(() => SystemInfo.deviceName);
        sys["deviceModel"] = (Func<string>)(() => SystemInfo.deviceModel);
        sys["processorCount"] = (Func<int>)(() => SystemInfo.processorCount);
        sys["systemMemoryMB"] = (Func<int>)(() => SystemInfo.systemMemorySize);

        sys["documentsPath"] = (Func<string>)(() =>
            Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments));
        sys["persistentPath"] = (Func<string>)(() => Application.persistentDataPath);
        sys["tempPath"] = (Func<string>)(() => Application.temporaryCachePath);

        sys["frameCount"] = (Func<int>)(() => Time.frameCount);
        sys["deltaTime"] = (Func<float>)(() => Time.deltaTime);

        script.Globals["System"] = sys;
    }
}
